patchwork-extensions-blockentity/src/main/java/net/patchworkmc/mixin/extensions/blockentity/MixinClientPlayNetworkHandler.java

```diff
@@ -0,0 +1,121 @@
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.At.Shift;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+import net.minecraftforge.common.extensions.IForgeTileEntity;
+
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.client.network.ClientPlayNetworkHandler;
+import net.minecraft.client.world.ClientWorld;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.packet.s2c.play.BlockEntityUpdateS2CPacket;
+import net.minecraft.network.packet.s2c.play.ChunkDataS2CPacket;
+import net.minecraft.util.math.BlockPos;
+
+import net.fabricmc.fabric.api.block.entity.BlockEntityClientSerializable;
+
+/**
+ * TODO: This Mixin can be less hacky if we can talk to the Fabric API team.
+ *
+ * <p>This Mixin implements {@link IForgeTileEntity#handleUpdateTag(CompoundTag)}
+ * and {@link IForgeTileEntity#onDataPacket(net.minecraft.network.ClientConnection, BlockEntityUpdateS2CPacket)}.
+ */
+@Mixin(ClientPlayNetworkHandler.class)
+public abstract class MixinClientPlayNetworkHandler {
+	///////////////////////////////////////////////////////////////
+	/// onChunkData -> IForgeTileEntity.handleUpdateTag
+	///////////////////////////////////////////////////////////////
+	@Unique
+	private static final ThreadLocal<CompoundTag> onChunkData_BETag = ThreadLocal.withInitial(() -> null);
+	@Unique
+	private static final String ClientWorld_getBlockEntity = "net/minecraft/client/world/ClientWorld.getBlockEntity(Lnet/minecraft/util/math/BlockPos;)Lnet/minecraft/block/entity/BlockEntity;";
+
+	@Inject(method = "onChunkData", locals = LocalCapture.CAPTURE_FAILHARD,
+			at = @At(value = "INVOKE", shift = Shift.BEFORE, target = ClientWorld_getBlockEntity))
+	private void onChunkData_CollectBETag(ChunkDataS2CPacket packet, CallbackInfo ci,
+			@SuppressWarnings("rawtypes") Iterator blockEntityTagListIterator, CompoundTag blockEntityTag, BlockPos blockPos) {
+		if (onBlockEntityUpdate_BlockEntity.get() != null) {
+			throw new IllegalStateException("State of ClientPlayNetworkHandler.onChunkData() is not clean, incompatible Mixins might be the cause!");
+		}
+
+		onChunkData_BETag.set(blockEntityTag);
+	}
+
+	@Redirect(method = "onChunkData", at = @At(value = "INVOKE", target = ClientWorld_getBlockEntity))
+	private BlockEntity onChunkData_world_getBlockEntity(ClientWorld clientWorld, BlockPos blockPos) {
+		CompoundTag blockEntityTag = onChunkData_BETag.get();
+		onChunkData_BETag.remove();
+
+		ClientPlayNetworkHandler me = (ClientPlayNetworkHandler) (Object) this;
+		BlockEntity blockEntity = me.getWorld().getBlockEntity(blockPos);
+
+		if (blockEntity != null && !(blockEntity instanceof BlockEntityClientSerializable)) {
```

Comment:
This looks like a hack. Would a marker interface in Patcher be helpful?

Reply:
explained in the previous comment